lib/query.js

```diff
@@ -57,3 +102,111 @@ export const getValue = node => {
+    if (isNumber(node)) {
+        return 0
+    } else if (isIdentifier(node) || isPolynomial(node)){
+        return 1
+    } else if (isPow(node)) {
+        return getValue(node.args[1])
+    } else if (isMul(node)){
+        return getExponent(node.args[1])
+    } else if (isNeg(node)) {
+        const variable = node.args[0]
+        return getExponent(variable.args[1])
+    } else {
+        return null
+    }
+}
+
+export const getCoefficient = (node) => {
+    if (isNumber(node)) {
+        return node
+    } else if (isIdentifier(node) || isPow(node)) {
+        return build.numberNode(1)
+    } else if (isNeg(node)) {
+        const result = build.applyNode('neg', [getCoefficient(node.args[0])])
+        result.wasMinus = node.wasMinus
+        return result
+    } else if (isMul(node)) {
+        const numbers = node.args.filter(arg => isNumber(arg) || isConstantFraction(arg))
+        if (numbers.length > 1) {
+            return build.applyNode('mul', numbers)
+        } else if (numbers.length > 0) {
+            return numbers[0]
+        } else {
+            return build.numberNode(1)
+        }
+    }
+}
+
+export const getVariableFactors = (node) => {
+    if (isVariableFactor(node)) {
+        return [node]
+    } else if (isMul(node)) {
+        return node.args.filter(isVariableFactor)
+    } else if (isNeg(node)) {
+        // TODO: figure out how to handle (x)(-y)(z)
+        return getVariableFactors(node.args[0])
+    } else {
+        // throw?
+    }
+}
+
+export const getVariableFactorName = (node) => {
+    if (isIdentifier(node)) {
+        return node.name
+    } else if (isPow(node)) {
+        return getVariableFactorName(node.args[0])
+    } else {
+        // throw?
+    }
+}
+
+export const sortVariables = (variables) =>
+    variables.sort(
+        (a, b) => getVariableFactorName(a) > getVariableFactorName(b))
+
+// TODO: fix the case of single polynomials ex. 2x^2
+export const getCoefficientsAndConstants = (node) => {
+    const coefficientMap = {}
+    const constants = []
+
+    node.args.forEach(arg => {
+        if (isNumber(arg) || isConstantFraction(arg)) {
+            constants.push(arg)
+        } else {
+            const sortedVariables = sortVariables(getVariableFactors(arg))
+
+            const coefficient = getCoefficient(arg)
+            const implicit = isImplicit(arg)
+
+            const key = sortedVariables.length > 1
+                  ? print(build.applyNode('mul', sortedVariables, {implicit}))
+                  : print(sortedVariables[0])
+
+            if (!(key in coefficientMap)) {
+                coefficientMap[key] = []
+            }
+
+            coefficientMap[key].push(coefficient)
+        }
+    })
+
+    return {coefficientMap, constants}
+}
+
+// returns true if two nodes have the same base variable
+// e.g. (x + 1)^1 and (x + 1)^3 , true
+// e.g. x^2, (x + 1)^2, false
+export const hasSameBase = (node1, node2) => {
+    return print(node1) == print(node2)
```

Comment:
Really we should have an `nodeEquals` function that traverse both nodes.  We have something that almost does this in `math-rules`.  I can add a `nodeEquals` function.